beets/ui/commands.py

```diff
@@ -801,6 +830,53 @@ def should_resume(self, path):
                            "was interrupted. Resume (Y/n)?"
                            .format(displayable_path(path)))
 
+    def _get_plugin_options(self, task):
+        """Get the extra options appended to the plugins to the ui prompt.
+
+        The `before_choose_candidate` event is sent to the plugins, with
+        session and task as its parameters. Plugins are responsible for
+        checking the right conditions and returning a list of `PromptChoice`s,
+        which is flattened and checked for conflicts.
+
+        Raises `ValueError` if two of the choices have the same short letter.
+
+        Returns a list of `PromptChoice`s.
+        """
+        # Send the before_choose_candidate event and flatten list.
+        extra_choices = list(chain(*plugins.send('before_choose_candidate',
+                                                 session=self, task=task)))
+        # Take into account default options, for duplicate checking.
+        all_choices = [PromptChoice(self, importer.action.SKIP, 's', 'Skip',
+                                    None),
+                       PromptChoice(self, importer.action.ASIS, 'u',
+                                    'Use as-is', None),
+                       PromptChoice(self, importer.action.TRACKS, 't',
+                                    'as Tracks', None),
+                       PromptChoice(self, importer.action.ALBUMS, 'g',
+                                    'Group albums', None),
+                       PromptChoice(self, importer.action.MANUAL, 'e',
+                                    'Enter search', None),
+                       PromptChoice(self, importer.action.MANUAL_ID, 'i',
+                                    'enter Id', None),
+                       PromptChoice(self, '', 'b', 'aBort', None)] +\
+            extra_choices
+
+        short_letters = [c.short for c in all_choices]
+        if len(short_letters) != len(set(short_letters)):
+            # Duplicate short letter has been found.
+            duplicates = [i for i, count in Counter(short_letters).items()
+                          if count > 1]
+            # Build informative message: 'x': classname:"long option",...
+            dup = {letter: ['%s:"%s"' % (type(c.plugin).__name__, c.long)
+                            for c in all_choices if c.short == letter]
+                   for letter in duplicates}
+            conflict_msg = '; '.join("'%s': %s" % (k, ', '.join(v))
+                                     for (k, v) in dup.iteritems())
+            raise ValueError("Prompt options have the same short letter\n%s" %
```

Comment:
This seems totally reasonable. My only question is whether we really need to have the `plugin` member for `PromptChoice`. It helps a bit with debuggability here, but plugin authors should be able to tell when they are responsible for creating a conflict. And if we want to have a uniform interface with built-in choices, then it will become decreasingly relevant. For example, "TerminalImportSession" is already a little weird in a diagnostic message.